BDArmory/Ammo/PooledBullet.cs

```diff
@@ -1603,19 +1603,20 @@ private bool ProximityAirDetonation(float distanceFromStart)
                 if (BDArmorySettings.VESSEL_RELATIVE_BULLET_CHECKS) //additional check if fast-moving vessel advances part bullet position/predicted pos during next FlightIntegrator tick. Possibly repalce OverlapSphere entirely with this method?
                 */ 
                 {
+                    Vector3 prevPos = currPosition - currentVelocity * TimeWarp.fixedDeltaTime; 
                     using (var loadedVessels = BDATargetManager.LoadedVessels.GetEnumerator())
                     {
                         while (loadedVessels.MoveNext())
                         {
                             if (loadedVessels.Current == null || !loadedVessels.Current.loaded) continue;
                             if (loadedVessels.Current == sourceVessel) continue;
                             float detRangeTime = TimeWarp.fixedDeltaTime * (detonationRange / (currentSpeed * TimeWarp.fixedDeltaTime)); //if detonation range > distance bullet moves in 1 frame, increase lok time to time it takes for bullet to move proxiRange
-                            timeToCPA = AIUtils.TimeToCPA(loadedVessels.Current, currPosition, currentVelocity, Vector3.zero, detRangeTime);
+                            timeToCPA = AIUtils.TimeToCPA(loadedVessels.Current, prevPos, currentVelocity, Vector3.zero, detRangeTime);
                             if (timeToCPA > detRangeTime) timeToCPA = 0;
                             if (timeToCPA > 0)
                             {
-                                Vector3 adjustedTgtPos = loadedVessels.Current.CoM + (currPosition - loadedVessels.Current.CoM).normalized * (loadedVessels.Current.GetRadius() + detonationRange);
-                                Vector3 CPA = AIUtils.PredictPosition(currPosition, currentVelocity, bulletDrop ? FlightGlobals.getGeeForceAtPosition(transform.position) : Vector3.zero, timeToCPA);
+                                Vector3 adjustedTgtPos = loadedVessels.Current.CoM + (prevPos - loadedVessels.Current.CoM).normalized * (loadedVessels.Current.GetRadius() + detonationRange);
+                                Vector3 CPA = AIUtils.PredictPosition(prevPos, currentVelocity, bulletDrop ? FlightGlobals.getGeeForceAtPosition(transform.position) : Vector3.zero, timeToCPA);
                                 if ((CPA - adjustedTgtPos).sqrMagnitude < detonationRange * detonationRange) //for head-on intercepts, t2CPA will likely be near 0; adjust to back when CPA just reaches detonationRange to give an appropriate currPos + currVel * time2CPA offset for visuals
                                 {
                                     timeToCPA -= ((detonationRange - (CPA - adjustedTgtPos).magnitude) / (currentSpeed * TimeWarp.fixedDeltaTime)) * TimeWarp.fixedDeltaTime;
```

Comment:
That `return detonate = ...` should probably actually continue until `detonate` is `true` in the unlikely event that there's another vessel that's also going to be within range.

Reply:
Thinking about this some more, I'm not sure of that upper bound for `detRangeTime`. It may want to be up to double that. I'll have to think some more about it.

Reply:
In testing your rewrite resulted in worse behavior, with detonation now occurring at best closer to the target (it doesn't seem to be taking target radius into account, at least in static tests?) and at worse behind the target (when in orbit) depending on shooter/target orientation, both before and after fixing the detonation/subprojectile spawn offset issue.